tools/wake-format/emitter.cpp

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright 2022 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Open Group Base Specifications Issue 7
+#define _XOPEN_SOURCE 700
+#define _POSIX_C_SOURCE 200809L
+
+#include "emitter.h"
+
+#include <cassert>
+
+#include "parser/parser.h"
+
+wcl::rope Emitter::newline(ctx_t ctx) {
+  wcl::rope_builder builder;
+
+  builder.append("⏎\n");
+  for (int i = 0; i < ctx.nest_level; i++) {
+    builder.append(space(space_per_indent));
+  }
+
+  return std::move(builder).build();
+}
+
+wcl::rope Emitter::space(uint8_t count) {
+  wcl::rope_builder builder;
+  for (uint8_t i = 0; i < count; i++) {
+    builder.append("·");
+  }
+  return std::move(builder).build();
+}
+
+wcl::optional<wcl::rope> Emitter::flat(ctx_t ctx, CSTElement node) {
+  return walk_node(ctx.flat(), node);
+}
+
+wcl::rope Emitter::try_flat(ctx_t ctx, CSTElement node) {
+  auto flat_node = flat(ctx, node);
+  if (flat_node && flat_node->size() + ctx.nest_level * space_per_indent < 80) {
+    return *flat_node;
+  }
+  auto full_node = walk_node(ctx, node);
+  assert(full_node);
+  return *full_node;
+}
+
+#define ASSERT_TOKEN(node, token) \
+  assert(!node.empty());          \
+  assert(node.id() == token);
+
+wcl::rope Emitter::layout(CST cst) {
+  ctx_t ctx;
+  return walk_top(ctx, cst.root()).concat(newline(ctx));
+}
+
+wcl::rope Emitter::walk_top(ctx_t ctx, CSTElement node) {
+  wcl::rope_builder builder;
+
+  for (CSTElement child = node.firstChildElement(); !child.empty(); child.nextSiblingElement()) {
```

Comment:
hmm...does it not matter what type of node this is? It seems to me like you should look at the top of `node` before traversing its children right?

Reply:
Eh yes, kinda but not this case. The follow up PR is much more explict though.

In this case,  this is only every called on `TOP` and `BLOCK` nodes. Both of which are just a list of other nodes, no specially handling required.

Reply:
I see. That makes more sense. Can we name this 'walk block' instead then? I think of `top` as just being a top level block anyway. Also comments saying which node/token types each walk function handles would be nice when they handle more than one.

Reply:
There is actually a subtle difference between walking `top` and `block` in the follow up PR "walk top" is just called `walk` and "walk block" is called `walk_block`.